spinalcordtoolbox/gui/centerline.py

```diff
@@ -193,14 +234,24 @@ def on_skip_slice(self):
 
     def on_toggle_mode(self):
         widget = self.sender()
-        if widget.mode in self._controller.MODES and widget.isChecked() and widget.mode != self._controller.mode:
-            self._controller.mode = widget.mode
-            self.update_status('Reset manual labels: Now in mode {}'.format(widget.mode))
-
-            self.sagittal_canvas.title(widget.sagittal_title)
-            self.sagittal_canvas.refresh()
-            self.axial_canvas.title(widget.axial_title)
-            self.axial_canvas.refresh()
+        # If needed in case user used keyboard shortcut (no widget.mode in this case)
+        if hasattr(widget, 'mode'):
+            if widget.mode == self._controller.mode:
+                # User clicked on the same button, so no further action is required.
+                return
+        else:
+            # If user selected keyboard shortcut, simply toggle mode without further checks.
+            if self._controller.mode == 'AUTO':
+                widget = self.custom_mode
+            else:
+                widget = self.auto_mode
+            widget.click()
+        self._controller.mode = widget.mode
+        self.update_status('Now in mode {}'.format(widget.mode))
```

Comment:
Hmm... this function effectively calls itself via `widget.click()`. The recursion makes this a little hard to understand on first read. I think it would be a little clearer if there were two separate functions. This also means you don't have to check for mode with `hasattr`.

```python
# in _init_controls()
custom_mode.toggled.connect(self.on_toggle_mode_click)
auto_mode.toggled.connect(self.on_toggle_mode_click)
QtWidgets.QShortcut(QtGui.QKeySequence('Ctrl+T'), self, self.on_toggle_mode_shortcut)

def on_toggle_mode_shortcut(self):
    # Keyboard shortcut pressed, so emulate clicking the button that would switch the mode
    if self._controller.mode == 'AUTO':
        self.custom_mode.click()
    else:
        self.auto_mode.click()   
    
def on_toggle_mode_click(self):
    widget = self.sender()
    self._controller.mode = widget.mode
    self.update_status('Now in mode {}'.format(widget.mode))
    # ...
```

Also, this check wasn't necessary:

```python
 if widget.mode == self._controller.mode:
      # User clicked on the same button, so no further action is required.
      return
```

Because if the modes are the same, then nothing will change anyway. :stuck_out_tongue:

Reply:
ah! i'm glad you have a suggestion there. I literally spent an hour banging my head on a wall with this "toggle" subtlety, and I felt dirty to dump a code which, despite it doing the job, is far from being elegant 😬

so, could you please also add that to the list in https://github.com/neuropoly/spinalcordtoolbox/issues/3178? thank you!

> Also, I don't think this check was necessary:

I was also wondering about this... i haven't tested, but i thought it would just address the case where the user clicks on the already-selected radio button: in this case we would rather have no action at all, instead of going through windows refreshing, message being re-displayed, etc. although again, i haven't tested that behaviour, so maybe that not a big deal (even not a minor deal, or no deal at all 🙂)